### PR TITLE
Fixes #12256 - removed explicit transactions

### PR DESCRIPTION
**Pull request by lzap**
Causing issues since core orchestration is not ready for them.

### CODE REVIEW

**Comment by iNecas** [comment]
I've tested the patch and I can confirm it fixes the issue with the post_commit happening while the orchestration fails: I've explicitely raise an error in `setTFTP` orchestration method, without this patch, the discovered host got restarted even when some previous orchestration code failed. With the patch, no restarted happened. I also can confirm the standard workflow with provisioning a discovered host works. I have not tested though all the cases where this PR touches the code.

Also, at least an issue should be raised on the foreman side about the nested transactions issue: this might cause other errors somewhere else then the discovery.

File: app/models/host/managed_extensions.rb [hunk 1]
```diff
@@ -34,6 +34,10 @@ def setReboot
     true
   end
 
+  def delReboot
```
Comment (ohadlevy) [inline app/models/host/managed_extensions.rb:40]: not related to this patch?

Comment (lzap) [inline app/models/host/managed_extensions.rb:40]: > not related tot his patch?

It is, it removes exception from logs about missing rollback method when
transaction fails.

## 

Later,
 Lukas #lzap Zapletal

**Comment by lzap** [comment]
Rebased @orrabin

**Comment by lzap** [comment]
Rebased again @orrabin

**Comment by lzap** [comment]
And .... rebased once again :-)

**Comment by dLobatog** [comment]
Merged as df4df986ce943e61cac2d71bd5af6e7a2816bb47, thanks @lzap!

**Comment by lzap** [comment]
@iNecas I am evaluating patches for backporting into Satellite 6.2 and it looks like you had some problems which were solved by this patch. What was it?

**Comment by iNecas** [comment]
The problem was, when some orchestration failed (such as setFTF), the host still got restarted even when the transaction failed. It's probably not that big deal, as if everything works, there are no rollbacks. The only issue here is, that there is a delay until the host gets restarted on rollback. The problem is conceptual: this means that if anyone wraps the `save` method with orchestration into transaction, they can expect similar issues as here, so the post_commit orchestration happens regardless success of failure of pre_commit orchestration.

**Comment by iNecas** [comment]
I've filled an issue against Foreman describing what the real problem is http://projects.theforeman.org/issues/14001

**Comment by lzap** [comment]
Ok, for the record, I am keeping this patch out of Satellite 6.2 as it's not blocker. It has been merged already and is available in Discovery 5.0.1 release.

**Comment by lzap** [comment]
For the record, this also fixes known interaction in Rails (< 4.1.6) between ActiveRecord::Base#becomes and ActiveRecord::Base#save!. The issue is that the save! call creates an UPDATE like this: 

```
UPDATE "hosts" SET ... WHERE "hosts"."type" IN ('Host::Managed') AND "hosts"."id" = 18  [...]
```

The WHERE clause there matches nothing, because the record in the database still has a type of Host::Discovered. An UPDATE on an empty set doesn't raise an error, so execution continues, and the transaction is not rolled back. This means, if you convert a host from discovered to managed and then save! it inside a transaction, all of the SQL statements that handle deleting the discovered attributes, adding entries to hosts/puppetclasses/operating systems/hostgroups etc succeed, and at the end you are left with a broken discovered host with no attributes, that can't even be deleted because of database constraints.

This is only an issue in Rails versions <= 4.1.5, it was fixed with this commit in Rails 4.1.6: https://github.com/rails/rails/commit/257d65ed2ceb14dca6b1319355ed4f8a73540498 . Here is the associated issue: https://github.com/rails/rails/issues/14785

https://bugzilla.redhat.com/show_bug.cgi?id=1357583

**Pull request closed without merging**